content/posts/gsoc-2023-dwarf/index.md

```diff
@@ -0,0 +1,73 @@
+title: "GSoC 2023 - Enhancing DWARF Support"
+date: "2023-09-06"
+summary: "In this article, I discuss my experience enhancing DWARF support in Rizin for Google Summer of Code 2023."
+tags: ["rizin", "dwarf", "gsoc"]
+ShowToc: false
+weight: 2
+---
+
+Hi! I'm [Billow](https://github.com/imbillow), and I had the privilege of participating in GSoC 2023, working on improving DWARF support for the Rizin project. In this blog post, I'm excited to share my journey, the challenges I faced, and my future plans for this project. Let's dive right in!
+
+Over the past few months, my primary focus has been on enhancing DWARF support within Rizin. DWARF (Debugging With Arbitrary Record Formats) is a crucial standard for debugging information in binary files. My work brings significant improvements, including the introduction of exprloc, compressed debug sections and composite variable storage.
+
+To showcase some of my achievements, I'm comparing the disassembly output obtained using the `pdf` command for the `write_fmt<Write>` function in the ELF file [dwarf_rust_bubble ↗](https://github.com/rizinorg/rizin-testbins/raw/master/elf/dwarf_rust_bubble) before and after my DWARF contributions were integrated. The enhanced output demonstrates Rizin's improved ability to parse DWARF debugging information and precisely locate variables.
+
+```
+[0x00005180]> pdf @ dbg.write_fmt_Write
+...
+┌ dbg.write_fmt<Write>();
+│           ; var int64_t var_28h @ stack - 0x28
+│           ; var int64_t var_18h @ stack - 0x18
+│           0x00010270      push  rbx                                  ; impls.rs:155 ; struct Result<(), std::io::error::Error> write_fmt<Write>(struct Box<Write> *self, struct Arguments fmt);
+...
+```
+
+```
+[0x00005180]> pdf @ dbg.write_fmt_Write
+...
+┌ struct Result<(), std::io::error::Error> write_fmt<Write>(struct Box<Write> *self, struct Arguments fmt)
+...
+│           ; arg struct Box<Write> *self @ rsi
+│           ; arg struct Arguments fmt @ ...
+│           0x00010270      push  rbx                                  ; impls.rs:155 ; struct Result<(), std::io::error::Error> write_fmt<Write>(struct Box<Write> *self, struct Arguments fmt)
+...
+```
+
+Another example, I'm comparing the disassembly output obtained using the `pdf` command for the `iterPreorder` function in the ELF file [dwarf_go_tree ↗](https://github.com/rizinorg/rizin-testbins/raw/master/elf/dwarf_go_tree). The `iterPreorder` function shows arguments and local variables precisely located on the stack. Most notably, the tree argument is represented as a composite variable spread across multiple stack locations. This new composite storage capability handles complex DWARF types. Overall, the improved output matches the original DWARF debugging data much more closely, rather than using generic `unknown` types. This showcases Rizin’s significantly upgraded DWARF parsing including features like composite variables and the immense value delivered to reverse engineers through my work.
```

Comment:
You can omit "I'm comparing" here, it's a soft tautology with the previous sentence.